src/hotspot/share/runtime/javaThread.cpp

```diff
@@ -556,27 +566,21 @@ bool JavaThread::is_interrupted(bool clear_interrupted) {
     assert(this == Thread::current(), "invariant");
     return false;
   }
+  JavaThread* current = JavaThread::current();
+  HandleMark hm(current);
+
+  Handle thread_h(current, vthread_or_thread());
+  ObjectLocker lock(Handle(current, java_lang_Thread::interrupt_lock(thread_h())), current);
```

Comment:
For this version then I assume you should limit it when its a virtual thread and when clear_interrupted is true.

Also need to think through if Object.wait will need to be changed as part of this.

Reply:
Thanks, I'm already working on it.

Reply:
Fixed as suggested by Alan.

Reply:
Use of `ObjectLocker` here will introduce a new pinning point for Loom. We have been removing as many uses of `ObjectLocker` as we can. I also think this will need to be moved back to Java code when the pinning currently inherent in calling `Object.wait` is addressed.

Reply:
Yes, and it may be that once Object.wait is implemented that we can remove the need to propagate the interrupt status (there are some TBDs here).

I think the change here is okay for now but we still have the choice of limiting the change to just JVMTI RawMonitorWait.

Reply:
Personally I'd prefer to see changes limited to just JVMTI `RawMonitorWait`. That minimises the risk of any unintended consequences from making the change.

Reply:
I've restored the `InterruptedException` catch in the `Object.wait`.
However, the fix in `JavaThread::is_interrupted()` also impacts the variations of `sleep_nanos()`.

Reply:
So the current changes do not limit this to just `RawMonitorWait`. I was expecting to only see additional code in `RawMonitorWait` that emulates what the Java code does to get the virtual and carrier thread interrupt states in sync, using the interruptLock.

Reply:
Okay, I can do this separation.
One minor concern is the `JavaThread::sleep_nanos(jlong nanos)` ?
It is used by the `JavaThread::sleep(jlong millis)` which in tern is not used much.

Reply:
Fixed now.